### ISSUE TITLE
[Yocto] gdb remote connection failing to load symbols

### ISSUE THREAD

**Issue by fdk17** [microsoft/vscode-cpptools#4780]
Type: Debugger
Trying to setup gdb to connect to gdbserver with a Yocto toolchain is not working.  The following errors are always shown:

```
Warning: Debuggee TargetArchitecture not detected, assuming x86_64.
=cmd-param-changed,param="pagination",value="off"
=tsv-created,name="trace_timestamp",initial="0"\n
0x00007ffff7fd5090 in ?? ()
ERROR: Command aborted. See the output window for additional details.
```

- OS and Version:  Ubuntu 18.04
- VS Code Version: 1.41.1
- C/C++ Extension Version: 0.26.2

`launch.json` configuration
```
        {
            "name": "(gdb) Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "/swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug/Xorg",
            "args": [],
            "stopAtEntry": false,
            "cwd": "/swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "setupCommands": [
                {
                    "text": "set sysroot /swdev/snc-dev/yocto-snc/build_poky/debugfs/",
                    "ignoreFailures": false
                },
            ],
            "miDebuggerServerAddress": "192.168.42.225:1323",
            "miDebuggerPath": "/swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/x86_64-pokysdk-linux/usr/bin/x86_64-poky-linux/x86_64-poky-linux-gdb",
            "logging": {
                "trace": false,
                "traceResponse": false,
                "engineLogging": false
            }
        }
```
Output from `engineLogging`:

```
1: (135) LaunchOptions<LocalLaunchOptions xmlns='http://schemas.microsoft.com/vstudio/MDDDebuggerOptions/2014'
1: (147) LaunchOptions  ExePath='/swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug/Xorg'
1: (148) LaunchOptions  WorkingDirectory='/swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug'
1: (148) LaunchOptions  ExeArguments=''
1: (148) LaunchOptions  MIMode='gdb'
1: (148) LaunchOptions  MIDebuggerPath='/swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/x86_64-pokysdk-linux/usr/bin/x86_64-poky-linux/x86_64-poky-linux-gdb'
1: (148) LaunchOptions  WaitDynamicLibLoad='false'
1: (148) LaunchOptions  MIDebuggerServerAddress='192.168.42.225:1323'
1: (148) LaunchOptions>
1: (149) LaunchOptions    <SetupCommands>
1: (149) LaunchOptions        <Command IgnoreFailures='false' Description=''>set sysroot /swdev/snc-dev/yocto-snc/build_poky/debugfs/</Command>
1: (149) LaunchOptions    </SetupCommands>
1: (149) LaunchOptions</LocalLaunchOptions>
1: (247) Starting: "/swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/x86_64-pokysdk-linux/usr/bin/x86_64-poky-linux/x86_64-poky-linux-gdb" --interpreter=mi
1: (264) DebuggerPid=4347
1: (297) ->=thread-group-added,id="i1"
1: (299) ->~"GNU gdb (GDB) 8.3.1\n"
1: (300) ->~"Copyright (C) 2019 Free Software Foundation, Inc.\n"
1: (300) ->~"License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\nThis is free software: you are free to change and redistribute it.\nThere is NO WARRANTY, to the extent permitted by law."
1: (300) ->~"\nType \"show copying\" and \"show warranty\" for details.\n"
1: (300) ->~"This GDB was configured as \"--host=x86_64-pokysdk-linux --target=x86_64-poky-linux\".\n"
1: (300) ->~"Type \"show configuration\" for configuration details.\n"
1: (301) ->~"For bug reporting instructions, please see:\n"
1: (301) ->~"<http://www.gnu.org/software/gdb/bugs/>.\n"
1: (301) ->~"Find the GDB manual and other documentation resources online at:\n    <http://www.gnu.org/software/gdb/documentation/>."
1: (301) ->~"\n\n"
1: (301) ->~"For help, type \"help\".\n"
1: (301) ->~"Type \"apropos word\" to search for commands related to \"word\".\n"
1: (301) ->(gdb)
1: (306) <-1001-gdb-set target-async on
1: (307) ->1001^done
1: (307) ->(gdb)
1: (309) 1001: elapsed time 3
1: (319) <-1002-interpreter-exec console "set sysroot /swdev/snc-dev/yocto-snc/build_poky/debugfs/"
1: (319) ->=cmd-param-changed,param="sysroot",value="/swdev/snc-dev/yocto-snc/build_poky/debugfs/"
1: (319) ->1002^done
1: (319) ->(gdb)
1: (320) 1002: elapsed time 0
1: (320) <-1003-interpreter-exec console "set pagination off"
1: (320) ->=cmd-param-changed,param="pagination",value="off"
1: (321) ->1003^done
1: (321) ->(gdb)
1: (321) 1003: elapsed time 0
1: (321) <-1004-gdb-set auto-solib-add on
1: (321) ->1004^done
1: (321) ->(gdb)
1: (321) 1004: elapsed time 0
1: (322) <-1005-gdb-set solib-search-path /swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug:
1: (322) ->1005^done
1: (322) ->(gdb)
1: (322) 1005: elapsed time 0
1: (322) <-1006-environment-cd /swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug
1: (322) ->1006^done
1: (323) ->(gdb)
1: (323) 1006: elapsed time 0
1: (323) <-1007-file-exec-and-symbols /swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug/Xorg
1: (437) ->1007^done
1: (438) ->(gdb)
1: (438) 1007: elapsed time 114
1: (438) <-1008-interpreter-exec console "show architecture"
1: (438) ->~"The target architecture is set automatically (currently i386:x86-64)\n"
1: (438) ->1008^done
1: (438) ->(gdb)
1: (438) 1008: elapsed time 0
1: (439) <-1009-break-insert -f main
1: (442) ->1009^done,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x0000000000043d10",func="main",file="../../xorg-server-1.20.5/dix/stubmain.c",fullname="/usr/src/debug/xserver-xorg/2_1.20.5-r0/build/dix/../../xorg-server-1.20.5/dix/stubmain.c",line="34",thread-groups=["i1"],times="0",original-location="main"}
1: (442) ->(gdb)
1: (446) 1009: elapsed time 6
1: (447) <-1010-target-select remote 192.168.42.225:1323
1: (453) ->=tsv-created,name="trace_timestamp",initial="0"\n
1: (453) ->=thread-group-started,id="i1",pid="579"
1: (453) ->=thread-created,id="1",group-id="i1"
1: (456) ->&"warning: Unable to find dynamic linker breakpoint function.\nGDB will be unable to debug shared library initializers\nand track explicitly loaded dynamic code."
1: (456) ->&"\n"
1: (458) <-1011-thread-info 1
1: (465) ->~"0x00007ffff7fd5090 in ?? ()\n"
1: (465) ->*stopped,frame={addr="0x00007ffff7fd5090",func="??",args=[],arch="i386:x86-64"},thread-id="1",stopped-threads="all",core="6"
1: (466) ->1010^connected
1: (466) ->(gdb)
1: (467) ->1011^done,threads=[{id="1",target-id="Thread 579.579",name="X",frame={level="0",addr="0x00007ffff7fd5090",func="??",args=[],arch="i386:x86-64"},state="stopped",core="6"}]
1: (467) ->(gdb)
1: (470) 1010: elapsed time 22
1: (471) 1011: elapsed time 13
1: (477) Send Event AD7EngineCreateEvent
1: (480) Send Event AD7ProgramCreateEvent
1: (524) Send Event AD7LoadCompleteEvent
=thread-group-added,id="i1"
GNU gdb (GDB) 8.3.1
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-pokysdk-linux --target=x86_64-poky-linux".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Warning: Debuggee TargetArchitecture not detected, assuming x86_64.
=cmd-param-changed,param="pagination",value="off"
=tsv-created,name="trace_timestamp",initial="0"\n
0x00007ffff7fd5090 in ?? ()
1: (538) <-1012-thread-info
1: (538) ->1012^done,threads=[{id="1",target-id="Thread 579.579",name="X",frame={level="0",addr="0x00007ffff7fd5090",func="??",args=[],arch="i386:x86-64"},state="stopped",core="6"}],current-thread-id="1"
1: (539) ->(gdb)
1: (539) 1012: elapsed time 1
1: (549) <-1013-stack-list-frames 0 1000
1: (551) ->1013^done,stack=[frame={level="0",addr="0x00007ffff7fd5090",func="??",arch="i386:x86-64"},frame={level="1",addr="0x0000000000000001",func="??",arch="i386:x86-64"},frame={level="2",addr="0x00007fffffffef1c",func="??",arch="i386:x86-64"},frame={level="3",addr="0x0000000000000000",func="??",arch="i386:x86-64"}]
1: (551) ->(gdb)
1: (551) 1013: elapsed time 2
1: (554) Send Event AD7ProcessInfoUpdatedEvent
1: (554) Send Event AD7ThreadCreateEvent
1: (562) <--exec-continue
1: (562) ->^running
1: (563) ->*running,thread-id="all"
1: (563) ->&"Warning:\n"
1: (563) ->&"Cannot insert breakpoint 1.\n"
1: (563) ->&"Cannot access memory at address 0x43d10\n"
1: (564) ->&"Cannot insert breakpoint -1.\n"
1: (564) ->&"Cannot access memory at address 0x43d20\n"
1: (564) ->&"\n"
1: (564) ->^error,msg="Command aborted."
1: (564) ->(gdb)
1: (566) Send Event AD7MessageEvent
ERROR: Command aborted. See the output window for additional details.
1: (568) <-1014-stack-info-frame
1: (568) ->1014^done,frame={level="0",addr="0x00007ffff7fd5090",func="??",arch="i386:x86-64"}
1: (568) ->(gdb)
1: (569) 1014: elapsed time 0
1: (569) <-1015-stack-list-frames 0 1000
1: (569) ->1015^done,stack=[frame={level="0",addr="0x00007ffff7fd5090",func="??",arch="i386:x86-64"},frame={level="1",addr="0x0000000000000001",func="??",arch="i386:x86-64"},frame={level="2",addr="0x00007fffffffef1c",func="??",arch="i386:x86-64"},frame={level="3",addr="0x0000000000000000",func="??",arch="i386:x86-64"}]
1: (569) ->(gdb)
1: (569) 1015: elapsed time 0
1: (570) Send Event AD7ExceptionEvent
Execute debugger commands using "-exec <command>", for example "-exec info registers" will list registers in use (when GDB is the debugger)
1: (642) <-1016-stack-list-arguments 0 0 0
1: (643) ->1016^done,stack-args=[frame={level="0",args=[]}]
1: (643) ->(gdb)
1: (644) 1016: elapsed time 1
```

The expected behavior (when running gdb at the command line):

```
(gdb) set sysroot /swdev/snc-dev/yocto-snc/build_poky/debugfs
(gdb) target remote 192.168.42.225:1323
Remote debugging using 192.168.42.225:1323
Reading symbols from /swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/core2-64-poky-linux/usr/bin/Xorg...
Reading symbols from /swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/core2-64-poky-linux/usr/bin/.debug/Xorg...
Reading symbols from /swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/core2-64-poky-linux/lib/ld-linux-x86-64.so.2...
Reading symbols from /swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/core2-64-poky-linux/lib/.debug/ld-2.30.so...
0x00007ffff7fd5090 in _start ()
   from /swdev/snc-dev/yocto-snc/opt/poky/3.0.1/sysroots/core2-64-poky-linux/lib/ld-linux-x86-64.so.2
```

I've tried using `additionalSOLibSearchPath` in conjunction with sysroot but nothing seems to work.

**Comment by pieandcakes** [comment]
I see a few problems in engine log you provided. symbol loading should be done automatically but it looks like the version of `gdb` that is being used in the `interpreter=mi` mode is not loading the symbols. 

I see the `solib-search-path` and `file-exec-and-symbols` paths being set:
```
1: (322) <-1005-gdb-set solib-search-path /swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug:
1: (322) ->1005^done
1: (322) ->(gdb)
1: (322) 1005: elapsed time 0
1: (322) <-1006-environment-cd /swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug
1: (322) ->1006^done
1: (323) ->(gdb)
1: (323) 1006: elapsed time 0
1: (323) <-1007-file-exec-and-symbols /swdev/snc-dev/yocto-snc/build_poky/debugfs/usr/bin/.debug/Xorg
1: (437) ->1007^done
```

but when the initial breakpoint is set, it looks like no symbols are loaded and gdb doesn't try and load any symbols:
```
1: (439) <-1009-break-insert -f main
1: (442) ->1009^done,bkpt={number="1",type="breakpoint",disp="keep",enabled="y",addr="0x0000000000043d10",func="main",file="../../xorg-server-1.20.5/dix/stubmain.c",fullname="/usr/src/debug/xserver-xorg/2_1.20.5-r0/build/dix/../../xorg-server-1.20.5/dix/stubmain.c",line="34",thread-groups=["i1"],times="0",original-location="main"}
1: (442) ->(gdb)
```

Also, when the `target-select remote` command is issued, I see this error:
```
1: (456) ->&"warning: Unable to find dynamic linker breakpoint function.\nGDB will be unable to debug shared library initializers\nand track explicitly loaded dynamic code."
1: (456) ->&"\n"
```

This final error below is why debugging terminates. 
```
1: (562) <--exec-continue
1: (562) ->^running
1: (563) ->*running,thread-id="all"
1: (563) ->&"Warning:\n"
1: (563) ->&"Cannot insert breakpoint 1.\n"
1: (563) ->&"Cannot access memory at address 0x43d10\n"
1: (564) ->&"Cannot insert breakpoint -1.\n"
1: (564) ->&"Cannot access memory at address 0x43d20\n"
1: (564) ->&"\n"
1: (564) ->^error,msg="Command aborted."
```

**Comment by pieandcakes** [comment]
I would expect to see `->=library-loaded` events coming from `gdb`. Can you see if you can run `info sharedlibrary` and see what is actually loaded? If you can't get it while running in VS Code, you can start `gdb` with the `--interpreter=mi` flag to start it in `mi` mode.

**Comment by Feserich** [comment]
I got the similar problem with `gdb` 7.12 (actually I use the QNX GDB and gdbserver a.k.a. pdebug). 
The problem is the order of the `setupCommands`. The symbols gets probably loaded, if you execute the `set solib-search-path` after the `run` command. This is also mentioned [here](https://stackoverflow.com/questions/8611194/debugging-shared-libraries-with-gdbserver/45252113#45252113). 
The reason is GDB can't map the addresses before the dynamic loader on the target was executed. I also read, that this problem especially occurs with ARM processors. 

With the vscode-cpptools debugger this is also working for me, if I manually execute `-exec sharedlibrary` in the Debug Console, after all setup commands are finished and the debugger stops at the main function. I also added `additionalSOLibSearchPath` to the launch.json. 

@pieandcakes Is it possible to add the `sharedlibrary` command after run in the initial setup of the debugger? To force the execution of the dynamic loader on the target.
